### PR TITLE
issue 4492: DependencyMap.concrete checks for unconditional dependencies

### LINKED ISSUE #4107: Multiple providers found for blas
on a fresh spack ae9a9e019a73cb951d4d2a2585ac71a53f351c81 I suddenly see the following issue:
```
$ spack install dealii@develop%gcc@5.4.0 ^openblas
==> Error: Multiple providers found for 'blas': ['intel-parallel-studio@professional.2017.2%gcc@5.4.0~all+daal~ilp64+ipp+mkl+mpi~newdtags~openmp+rpath+shared+tools arch=linux-ubuntu16-x86_64', 'openblas@0.2.19%gcc@5.4.0~openmp+pic+shared arch=linux-ubuntu16-x86_64']
```
Yes, there are multiple providers for `blas`, but I don't understand why is it a problem now... Note that the problem is there even though there is `^openblas`.

p.s. this is without any `packages.yaml` file, only `compilers.yaml`

---

**TLDR**: currently a workaround is to specify both `mpi` and `lapack` providers, that is do
```
spack install dealii@develop%gcc@5.4.0+mpi ^openmpi ^openblas
```
---
MWE of the problem to be added to the unit tests is https://github.com/LLNL/spack/issues/4107#issuecomment-298889480

### LINKED ISSUE #4492: Virtual dependencies incorrectly affecting the hash
@tgamblin @scheibelp @alalazo @davydden This is a really weird one that I've never seen before. 

From a fresh clone of Spack, with no `packages.yaml`, I'm seeing the following behavior. If I don't specify an MPI provider and let it pick up the default (`openmpi`), I see:
```
$ spack spec -L hdf5+mpi
...
tgprd6nwfgifpaff4a7dvcv3tynuaowd  hdf5@1.10.1%gcc@6.3.1+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-fedora25-x86_64 
hj3i67mhesgu3x5ymwpj67m4nbq6d3c2      ^openmpi@2.1.1%gcc@6.3.1~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple+vt arch=linux-fedora25-x86_64 
eva5cqpkbjggxvpp23ycr62yiocie64g          ^hwloc@1.11.7%gcc@6.3.1~cuda+libxml2+pci arch=linux-fedora25-x86_64 
reafmcemqbmvzqzyiitmyfcdc4ia7tzl              ^libpciaccess@0.13.5%gcc@6.3.1 arch=linux-fedora25-x86_64 
4w4hvmepes4p75z37og3dn3gbzjpnwho                  ^libtool@2.4.6%gcc@6.3.1 arch=linux-fedora25-x86_64 
zcwwrrt7od3rrhxb5du2iybu65jlrlbw                      ^m4@1.4.18%gcc@6.3.1+sigsegv arch=linux-fedora25-x86_64 
625n6uiwuqes737tignpxu3snoibyc43                          ^libsigsegv@2.11%gcc@6.3.1 arch=linux-fedora25-x86_64 
ywniw4sq7xd3fzz6c4c7jrhqj5htwobl                  ^pkg-config@0.29.2%gcc@6.3.1+internal_glib arch=linux-fedora25-x86_64 
lnv6kl5fucdneeh7h7ickta4vdis7u4v                  ^util-macros@1.19.1%gcc@6.3.1 arch=linux-fedora25-x86_64 
ggtx25lpnfriqn4n35fsgquiug6uju7t              ^libxml2@2.9.4%gcc@6.3.1~python arch=linux-fedora25-x86_64 
yw6ju7g3kyc27eonzrrvu73cdpiy5hzq                  ^xz@5.2.3%gcc@6.3.1 arch=linux-fedora25-x86_64 
jcukpuaeyykzu647duoufe2bdrgru5ko                  ^zlib@1.2.11%gcc@6.3.1+pic+shared arch=linux-fedora25-x86_64 
```
If I do specify the MPI provider and explicitly tell it to use `openmpi`, I see:
```
$ spack spec -L hdf5+mpi^openmpi
...
ohnyu6ch2ewnkf2dnd32qrb6t2hlopgj  hdf5@1.10.1%gcc@6.3.1+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-fedora25-x86_64 
yyiri5gxf7jtu2c3vrhj6on4wz3jf2jh      ^openmpi@2.1.1%gcc@6.3.1~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple+vt arch=linux-fedora25-x86_64 
eva5cqpkbjggxvpp23ycr62yiocie64g          ^hwloc@1.11.7%gcc@6.3.1~cuda+libxml2+pci arch=linux-fedora25-x86_64 
reafmcemqbmvzqzyiitmyfcdc4ia7tzl              ^libpciaccess@0.13.5%gcc@6.3.1 arch=linux-fedora25-x86_64 
4w4hvmepes4p75z37og3dn3gbzjpnwho                  ^libtool@2.4.6%gcc@6.3.1 arch=linux-fedora25-x86_64 
zcwwrrt7od3rrhxb5du2iybu65jlrlbw                      ^m4@1.4.18%gcc@6.3.1+sigsegv arch=linux-fedora25-x86_64 
625n6uiwuqes737tignpxu3snoibyc43                          ^libsigsegv@2.11%gcc@6.3.1 arch=linux-fedora25-x86_64 
ywniw4sq7xd3fzz6c4c7jrhqj5htwobl                  ^pkg-config@0.29.2%gcc@6.3.1+internal_glib arch=linux-fedora25-x86_64 
lnv6kl5fucdneeh7h7ickta4vdis7u4v                  ^util-macros@1.19.1%gcc@6.3.1 arch=linux-fedora25-x86_64 
ggtx25lpnfriqn4n35fsgquiug6uju7t              ^libxml2@2.9.4%gcc@6.3.1~python arch=linux-fedora25-x86_64 
yw6ju7g3kyc27eonzrrvu73cdpiy5hzq                  ^xz@5.2.3%gcc@6.3.1 arch=linux-fedora25-x86_64 
jcukpuaeyykzu647duoufe2bdrgru5ko                  ^zlib@1.2.11%gcc@6.3.1+pic+shared arch=linux-fedora25-x86_64 
```
As you can see, the packages to be installed are identical. However, the hash for `openmpi` (and consequently `hdf5`) changed!

### PR DESCRIPTION
**Pull request by alalazo**
Fixes #4107 
Fixes #4492 

##### Modifications:

 - [x] added test to reproduce the issue
 - [x] fixed `DependencyMap.concrete`

### CODE REVIEW

**Comment by alalazo** [comment]
I am starting the PR early as I want to see the test xfailing on Travis before committing the fix to it.

**Comment by alalazo** [comment]
@becker33 @tgamblin 

With an implementaton of `DependencyMap.concrete` like:
```python
    @property
    def concrete(self):

        # Check if the dependencies we store are concrete and of the correct type
        dependencies_are_concrete = all(
            (d.spec.concrete and d.deptypes) for d in self.values()
        )

        if not dependencies_are_concrete:
            return False

        # Now check we have every dependency we need
        pkg = self.owner.package
        for dependency in pkg.dependencies.values():
            # Check that for every condition we satisfy we satisfy also
            # the associated constraint
            for condition, constraint in dependency.items():
                # WARNING: This part relies on the fact that dependencies are
                # the last item to be checked when computing if a spec is
                # concrete. This means that we are ensured that all variants,
                # versions, compilers, etc. are there with their final value
                # when we call 'Spec.satisfies(..., strict=True)'
                satisfy_condition = self.owner.satisfies(
                    condition, strict=True
                )
                satisfy_constraint = self.owner.satisfies(
                    constraint, strict=True
                )
                if satisfy_condition and not satisfy_constraint:
                    return False

        return True
```
which seems logically correct to me, I am in an infinite recursion loop because:
```
Spec.satisfies -> Spec.concrete -> DependencyMap.concrete -> Spec.satisfies -> ...
```

Any advice how to sort this out without complicating the implementation is welcome.

**Comment by alalazo** [comment]
A few sparse remarks, as I have the code still fresh in my mind:

- `DependencyMap` is used also for dependents. In that case the method `concrete` does not make much sense. Should we differentiate the types of the two objects then?

- #4129 is also related to MV variants, I think we should try to merge the two PRs at the same time (and check how the two interact together)

- As I stated in the commit comment, it is possible to forge a spec which will trigger exactly the same problem as before. This spec must have dependencies that are all conditionals. We may also have a case in which we miss a conditional dependency and we don't notice with the current implementation. The problem trying to do the concreteness check in a "logically correct" way is that we will trigger an infinite recursion (see my comment above)

Well, I will be waiting for people to stress test this and uncover as many problem as possible. :smile:

File: lib/spack/spack/spec.py [hunk 1]
```diff
@@ -678,10 +678,72 @@ class DependencyMap(HashableMap):
...
+        # Owner Spec for the current map
+        self.owner = owner
+
     @property
     def concrete(self):
-        return all((d.spec.concrete and d.deptypes)
-                   for d in self.values())
+
+        # Check if the dependencies are concrete and of the correct type
+        dependencies_are_concrete = all(
```
Comment (davydden) [inline lib/spack/spack/spec.py]: since there is a single place where you use this variable, you might as well do it in place.

File: lib/spack/spack/spec.py [hunk 2]
```diff
@@ -678,10 +678,72 @@ class DependencyMap(HashableMap):
...
+            (d.spec.concrete and d.deptypes) for d in self.values()
+        )
+
+        if not dependencies_are_concrete:
+            return False
+
+        # If we are dealing with an external spec, it clearly has no
+        # dependencies managed by spack
+        if self.owner.external:
```
Comment (davydden) [inline lib/spack/spack/spec.py]: a test would be good to have. I know it was not there from the beginning...

File: lib/spack/spack/spec.py [hunk 3]
```diff
@@ -678,10 +678,72 @@ class DependencyMap(HashableMap):
...
+            return True
+
+        # Now check we have every dependency we need
+        pkg = self.owner.package
+        for dependency in pkg.dependencies.values():
+            # Check that for every condition we satisfy we satisfy also
```
Comment (davydden) [inline lib/spack/spack/spec.py]: > we satisfy we satisfy

File: lib/spack/spack/spec.py [hunk 4]
```diff
@@ -678,10 +678,72 @@ class DependencyMap(HashableMap):
...
+            # the associated constraint
+            for condition, constraint in dependency.items():
+                # WARNING: This part relies on the fact that dependencies are
```
Comment (davydden) [inline lib/spack/spack/spec.py]: can this be tested via regression test? So that somebody do not accidentally re-order things in checking if spec is conrete?

**Comment by becker33** [comment]
@alalazo I think the best way to avoid the recursion is to create a property "locally_concrete" that checks everything except for the dependencies.

There are two places in `Spec.satisfies()` that can call `Spec.concrete()`. One of them I believe can be replaced with `locally_concrete`, and the other cannot be reached from `DependencyMap.concrete()`

The one that I think can be replaced is for testing multi-valued variants, so you may be in a better position than I am to think through that.

The one that cannot be reached from `DependencyMap.concrete()` is in the case where `other` is concrete. Since the condition spec can never be concrete, we won't see a loop in that case.

File: lib/spack/spack/spec.py (hunk 1 above)
Comment (alalazo) [inline lib/spack/spack/spec.py]: I did it on purpose to give a sensible name that reflects the semantic of the check. It's also true that I put a comment saying basically the same thing, so maybe I can remove one of the two.

File: lib/spack/spack/spec.py (hunk 2 above)
Comment (alalazo) [inline lib/spack/spack/spec.py]: Do you mean a test on an external spec? To be honest I didn't think of this case immediately. We have a test that fails if you remove this check, and I added the conditional to fix that. These two lines should be already covered by at least one test :smile:

Comment (davydden) [inline lib/spack/spack/spec.py]: If there's a test that fails without this line, it's good enough

File: lib/spack/spack/spec.py (hunk 1 above)
Comment (davydden) [inline lib/spack/spack/spec.py]: Fair enough, I would vote for keeping the comment, but I don't have a strong oppinion on that,

File: lib/spack/spack/spec.py (hunk 4 above)
Comment (alalazo) [inline lib/spack/spack/spec.py]: I can't think of any way to enforce this. I mean: should anybody reorder things I am pretty sure he will see a lot of failures in the tests. I put comments in both `Spec.concrete` and `DependencyMap.concrete` to make the cause of the failures stand out.

**Comment by alalazo** [comment]
@becker33 
> Does this make sense to you?

Yes, I think it's a good idea to try to relax the request on `self.concrete`. I'll try tomorrow to work out the details, but out of the blue I think I'll need something like "variants are already concrete everywhere", more than something like "locally concrete".

**Comment by becker33** [comment]
What about adding a flag to `DependencyMap.concrete()` that reverts it to the old behavior? Then `Spec.satisfies()` could pass a flag through `Spec.concrete()` that changes the behavior of `DependencyMap.concrete()` in that context.

Possible downside is this might require making these methods no longer be properties.

**Comment by alalazo** [comment]
@tgamblin @becker33 I am trying different solutions offline to relax the constraint and remove the FIXMEs in the code. I think it may take a while, because all I obtained in the last 2 hours is to have infinite recursion with a longer chain of calls. If it is enough as a workaround, and you don't see potential issues arising from this PR maybe we can merge it and open an issue to find a way to generalize the check on constraints for conditional dependencies.

> What about adding a flag to DependencyMap.concrete() that reverts it to the old behavior? Then Spec.satisfies() could pass a flag through Spec.concrete() that changes the behavior of DependencyMap.concrete() in that context.

I don't like the idea of adding a flag to `DependencyMap.concrete`, and vary its behavior depending on the context. My concern is about the semantic of the functions taking part in the concretization process. 

It's already quite complicated (at least for me) to think that `satisfies` means "could satisfy" during concretization and "can satisfy" afterwards. Making `DependencyMap.concrete` behave like "what I have is concrete, but it might not be all the dependencies I need" in only some contexts is making things more complicated, whereas I think we should try to simplify them.
